File: python/701.py
"""Project Euler Problem 701: Random connected area.

Find the expected maximum area of an orthogonally contiguous group of black
cells, if each cell in an N x N square is randomly colored white or black.

We build up board configurations, starting from the top left corner and going
in reading order, handling whether the current cell is white or black.

We memoize the configurations by only keeping track of the following state:
for each of the previous N cells, which are connected to which (the "profile",
which is 0 for each cell, and a distinct positive integer starting at 1 for
each set of black cells that are connected), the areas of the contiguous
regions containing those cells, and the total maximum area.

For handling the cell being white, we need only shift the profile and areas
left by 1. For handling the cell being black, we need to combine the groups
containing the cell above and to the left, if both are black, and sum their
areas if they are in different groups.

We store the number of configurations with each of the states, and at the end
multiply each maximum area by the corresponding probability (which is the
number of configurations for that state divided by 2^{N²}).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Tuple


@dataclass(frozen=True)
class State:
    """State representing a board configuration profile."""

    profile: Tuple[int, ...]
    areas: Tuple[int, ...]
    max_area: int


def canonicalize(profile: list[int], n: int) -> tuple[int, ...]:
    """Canonicalize a profile by remapping group IDs."""
    mapping: list[int] = [0] * (n + 1)
    new_profile: list[int] = [0] * n
    index = 0
    for i in range(n):
        num = profile[i]
        if num > 0 and mapping[num] == 0:
            index += 1
            mapping[num] = index
        new_profile[i] = mapping[num]
    return tuple(new_profile)


def solve() -> float:
    """Solve Problem 701."""
    n = 7
    states: Dict[State, int] = {
        State(tuple([0] * n), tuple([0] * n), 0): 1
    }

    for row in range(n):
        for col in range(n):
            new_states: Dict[State, int] = {}
            for state, count in states.items():
                # Handle white cell
                new_profile = [state.profile[i] for i in range(1, n)]
                new_profile.append(0)
                new_areas = [state.areas[i] for i in range(1, n)]
                new_areas.append(0)
                canonical_profile = canonicalize(new_profile, n)
                new_state = State(canonical_profile, tuple(new_areas), state.max_area)
                new_states[new_state] = new_states.get(new_state, 0) + count

                # Handle black cell
                new_profile = [state.profile[i] for i in range(1, n)]
                new_profile.append(0)
                new_areas = [state.areas[i] for i in range(1, n)]
                new_areas.append(0)
                new_area = (
                    1
                    + state.areas[0]
                    + (
                        state.areas[n - 1]
                        if state.profile[0] != state.profile[n - 1] and col > 0
                        else 0
                    )
                )
                for i in range(n):
                    if (
                        new_profile[i] > 0
                        and (
                            new_profile[i] == state.profile[0]
                            or (new_profile[i] == state.profile[n - 1] and col > 0)
                        )
                    ) or i == n - 1:
                        new_profile[i] = n
                        new_areas[i] = new_area
                canonical_profile = canonicalize(new_profile, n)
                new_state = State(
                    canonical_profile,
                    tuple(new_areas),
                    max(state.max_area, new_area),
                )
                new_states[new_state] = new_states.get(new_state, 0) + count

            states = new_states

    ans = 0.0
    for state, count in states.items():
        ans += count * state.max_area
    ans /= 2 ** (n * n)
    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(f"{result:.8f}")
    return int(result * 1e8)  # Return as integer for comparison


if __name__ == "__main__":
    main()

File: c/701.c
/*
 * Project Euler Problem 701: Random Connected Area.
 *
 * Find the expected maximum area of an orthogonally contiguous group of black
 * cells, if each cell in an N x N square is randomly colored white or black.
 *
 * Uses state-based DP: for each of the previous N cells, track which are
 * connected (profile), the areas of the contiguous regions, and the total
 * maximum area. We canonicalize profiles to reduce the state space.
 *
 * State: profile[N] (canonicalized group IDs), areas[N], max_area
 * We use a hash map of states -> counts.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NN 7
#define MAX_AREA 50  /* max possible area per group */
/* Hash table size - must be power of 2 */
#define HT_SIZE (1 << 22)
#define HT_MASK (HT_SIZE - 1)

typedef struct {
    unsigned char profile[NN];
    unsigned char areas[NN];
    unsigned char max_area;
} State;

typedef struct Entry {
    State state;
    long long count;
    struct Entry *next;
} Entry;

static Entry *table[HT_SIZE];
static Entry *new_table[HT_SIZE];

/* Pool allocator for entries */
#define POOL_BLOCK 1000000
static Entry *pool = NULL;
static int pool_idx = POOL_BLOCK;

Entry *alloc_entry(void) {
    if (pool_idx >= POOL_BLOCK) {
        pool = (Entry *)malloc(POOL_BLOCK * sizeof(Entry));
        pool_idx = 0;
    }
    return &pool[pool_idx++];
}

unsigned int hash_state(const State *s) {
    unsigned int h = 0;
    for (int i = 0; i < NN; i++) {
        h = h * 31 + s->profile[i];
        h = h * 31 + s->areas[i];
    }
    h = h * 31 + s->max_area;
    return h;
}

int state_eq(const State *a, const State *b) {
    return memcmp(a, b, sizeof(State)) == 0;
}

void ht_add(Entry **ht, const State *s, long long count) {
    unsigned int h = hash_state(s) & HT_MASK;
    Entry *e = ht[h];
    while (e) {
        if (state_eq(&e->state, s)) {
            e->count += count;
            return;
        }
        e = e->next;
    }
    e = alloc_entry();
    e->state = *s;
    e->count = count;
    e->next = ht[h];
    ht[h] = e;
}

void canonicalize(unsigned char *profile, unsigned char *out) {
    unsigned char mapping[NN + 1];
    memset(mapping, 0, sizeof(mapping));
    int idx = 0;
    for (int i = 0; i < NN; i++) {
        int num = profile[i];
        if (num > 0 && mapping[num] == 0) {
            idx++;
            mapping[num] = idx;
        }
        out[i] = mapping[num];
    }
}

int main() {
    memset(table, 0, sizeof(table));

    State init;
    memset(&init, 0, sizeof(init));
    ht_add(table, &init, 1);

    for (int row = 0; row < NN; row++) {
        for (int col = 0; col < NN; col++) {
            memset(new_table, 0, sizeof(new_table));
            /* Reset pool for new iteration */
            pool = NULL;
            pool_idx = POOL_BLOCK;

            for (int h = 0; h < HT_SIZE; h++) {
                Entry *e = table[h];
                while (e) {
                    State *st = &e->state;
                    long long count = e->count;

                    /* Handle white cell */
                    {
                        State ns;
                        unsigned char new_profile[NN], new_areas[NN];
                        for (int i = 0; i < NN - 1; i++) {
                            new_profile[i] = st->profile[i + 1];
                            new_areas[i] = st->areas[i + 1];
                        }
                        new_profile[NN - 1] = 0;
                        new_areas[NN - 1] = 0;
                        canonicalize(new_profile, ns.profile);
                        memcpy(ns.areas, new_areas, NN);
                        ns.max_area = st->max_area;
                        ht_add(new_table, &ns, count);
                    }

                    /* Handle black cell */
                    {
                        State ns;
                        unsigned char new_profile[NN], new_areas[NN];
                        for (int i = 0; i < NN - 1; i++) {
                            new_profile[i] = st->profile[i + 1];
                            new_areas[i] = st->areas[i + 1];
                        }
                        new_profile[NN - 1] = 0;
                        new_areas[NN - 1] = 0;

                        int above_group = st->profile[0];
                        int left_group = (col > 0) ? st->profile[NN - 1] : 0;

                        int new_area = 1 + st->areas[0];
                        if (above_group != left_group && col > 0) {
                            new_area += st->areas[NN - 1];
                        }

                        for (int i = 0; i < NN; i++) {
                            if ((new_profile[i] > 0 &&
                                 (new_profile[i] == above_group ||
                                  (new_profile[i] == left_group && col > 0)))
                                || i == NN - 1) {
                                new_profile[i] = NN;
                                new_areas[i] = new_area;
                            }
                        }

                        canonicalize(new_profile, ns.profile);
                        memcpy(ns.areas, new_areas, NN);
                        ns.max_area = st->max_area;
                        if (new_area > ns.max_area) ns.max_area = new_area;
                        ht_add(new_table, &ns, count);
                    }

                    e = e->next;
                }
            }

            /* Swap tables */
            memcpy(table, new_table, sizeof(table));
        }
    }

    /* Compute answer */
    double ans = 0.0;
    for (int h = 0; h < HT_SIZE; h++) {
        Entry *e = table[h];
        while (e) {
            ans += (double)e->count * e->state.max_area;
            e = e->next;
        }
    }

    /* Divide by 2^(N*N) */
    double total = 1.0;
    for (int i = 0; i < NN * NN; i++) total *= 2.0;
    ans /= total;

    printf("%.8f\n", ans);
    return 0;
}
